### DISCUSSION TITLE
GUI: missing drive types

### DISCUSSION THREAD

**Discussion by chenghuzi** [jonescompneurolab/hnn-core#490] [GSoC]
In our current GUI code, we have Rhythmic, Poisson and Evoked drives. And if load drives from `default.json` we got 4 types of them:
![image](https://user-images.githubusercontent.com/11160442/176987953-5790d2a4-bde8-4bb1-b132-c805f64549dd.png)

However, if we use `net = Network(params, add_drives_from_params=True)`, there are several other types of drives:
![image](https://user-images.githubusercontent.com/11160442/176987968-c82afd67-4c45-4929-b052-e9fe914f1ac6.png)

And the current GUI, with four drives, generates:
![image](https://user-images.githubusercontent.com/11160442/176987912-622577d8-0896-45e7-88f4-f76a90a9c989.png)

while the latter generates:
![image](https://user-images.githubusercontent.com/11160442/176988008-1d0ce794-7291-4b81-908e-3f697b96fc65.png)

**Comment by jasmainak** [comment]
The extra drives are to keep backwards compatibility with the old GUI. I think you can probably disable them by doing `legacy_mode=False`. @cjayb @rythorpe any opinions here?

**Reply by chenghuzi** [reply]
With 
```Python
import os.path as op

import hnn_core
import matplotlib.pyplot as plt
from hnn_core import (JoblibBackend, MPIBackend, jones_2009_model,
                      simulate_dipole, read_params, Network)
from hnn_core.viz import plot_dipole

hnn_core_root = op.join(op.dirname(hnn_core.__file__))
params_fname = op.join(hnn_core_root, 'param', 'default.json')
params = read_params(params_fname)

# net = jones_2009_model()
net = Network(params, add_drives_from_params=True, legacy_mode=True)
```
and inserting prints to `drives.py`
```Python
def _add_drives_from_params(net):
    drive_specs = _extract_drive_specs_from_hnn_params(
        net._params, list(net.cell_types.keys()))
    bias_specs = _extract_bias_specs_from_hnn_params(
        net._params, list(net.cell_types.keys()))

    for drive_name in sorted(drive_specs.keys()):  # order matters
        specs = drive_specs[drive_name]
        print(specs['type'], drive_name)
```
we have  as before

![image](https://user-images.githubusercontent.com/11160442/177192676-e3c6e358-6bf7-40f7-92b7-c0e9f11a1313.png)

However, if I set `legacy_mode=False` then I get error like this:
![image](https://user-images.githubusercontent.com/11160442/177192827-1efaaacb-d5d6-44e0-bb6c-afbe8ebba2c6.png)

**Reply by jasmainak** [reply]
arghgh ... what if you use `jones_2009_model(add_drives_from_params)`? This is why zen of Python says:

> There should be one-- and preferably only one --obvious way to do it.

I don't think `Network(add_drives_from_params)` is really meant to be used other than for replicating the old GUI results.

**Reply by chenghuzi** [reply]
So you mean `jones_2009_model ` is actually a general constructor for all possible network models? If so that will be much easier.

**Reply by chenghuzi** [reply]
Never mind. You've answered in another thread

**Comment by rythorpe** [comment]
I'm not sure I understand what's going on here. Why there is a difference between hnn-core and the new GUI when loading `default.json`, since they both use the same underlying code when handling parameter files?

**Reply by jasmainak** [reply]
Remember that the GUI always adds some "ghost" drives even if there are no spiking events associated with them. We removed this option in hnn-core over time.

**Reply by rythorpe** [reply]
Hmm, I still don't understand why the new GUI is doing this. Maybe we can discuss this in our next meeting?

**Reply by chenghuzi** [reply]
> Hmm, I still don't understand why the new GUI is doing this. Maybe we can discuss this in our next meeting?

It's the same now.

<img width="1459" alt="image" src="https://user-images.githubusercontent.com/11160442/178094594-37faea52-752d-48da-8428-565d9ef3bf8a.png">
Previously it was different as I followed the design that for certain drives we use the `tstop` value for simulation to overwrite the one from the parameter file. So this actually allowed some drives to invoke spikes, though in parameter they're disabled by setting `tstop` to 0.

**Comment by chenghuzi** [comment]
For all possible drive types, this what I have now from loading `default.json`:
<img width="389" alt="image" src="https://user-images.githubusercontent.com/11160442/178094693-847d7d86-5ec1-404b-827f-ce3e9d383003.png">

Please let me know if this is all of them @jasmainak @rythorpe . I checked `drives.py` and noticed that some drive names actually share the same method under the hood to add drives so I just slightly changed the logic in `gui.py` to fit this.

**Reply by jasmainak** [reply]
I think it's all of them!

**Reply by rythorpe** [reply]
Great, so just to confirm @chenghuzi, loading and simulating `default.json` using the hnn-core API (i.e., via scripting) vs the GUI now produce exactly the same number/type of drives as well as current dipole output?

**Reply by chenghuzi** [reply]
> Great, so just to confirm @chenghuzi, loading and simulating `default.json` using the hnn-core API (i.e., via scripting) vs the GUI now produce exactly the same number/type of drives as well as current dipole output?

yes
